### PR TITLE
kvserver: fail stale ConfChange when rejected by raft

### LINKED ISSUE #104709: roachtest: failover/chaos/read-write failed
roachtest.failover/chaos/read-write [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/10484309?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/10484309?buildTab=artifacts#/failover/chaos/read-write) on master @ [6e38f676273d08b3a111213d6c9f7945629228d7](https://github.com/cockroachdb/cockroach/commits/6e38f676273d08b3a111213d6c9f7945629228d7):

```
test artifacts and logs in: /artifacts/failover/chaos/read-write/run_1
(assertions.go:333).Fail: 
	Error Trace:	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/failover.go:1379
	            				github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/failover.go:288
	            				main/pkg/cmd/roachtest/monitor.go:105
	            				golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:75
	            				GOROOT/src/runtime/asm_amd64.s:1594
	Error:      	Received unexpected error:
	            	pq: replica unavailable: (n4,s4):11 unable to serve request to r792:/Table/106/1/-3{22495525764153008-04067210006201408} [(n4,s4):11, (n8,s8):10, (n3,s3):9, (n7,s7):6, (n6,s6):7, next=12, gen=76, sticky=9223372036.854775807,2147483647]: closed timestamp: 1686489678.462266747,0 (2023-06-11 13:21:18); raft status: {"id":"b","term":14,"vote":"b","commit":1205,"lead":"b","raftState":"StateLeader","applied":1205,"progress":{"6":{"match":1205,"next":1206,"state":"StateReplicate"},"7":{"match":1205,"next":1206,"state":"StateReplicate"},"9":{"match":1205,"next":1206,"state":"StateReplicate"},"a":{"match":1205,"next":1206,"state":"StateReplicate"},"b":{"match":1205,"next":1206,"state":"StateReplicate"}},"leadtransferee":"0"}: encountered poisoned latch /Local/Range/Table/106/1/-322495525764153008/RangeDescriptor@0,0
	Test:       	failover/chaos/read-write
(require.go:1360).NoError: FailNow called
(monitor.go:137).Wait: monitor failure: monitor task failed: t.Fatal() was called
(cluster.go:2276).Run: cluster.RunE: context canceled
(cluster.go:2276).Run: cluster.RunE: context canceled
(cluster.go:2276).Run: cluster.RunE: context canceled
(cluster.go:2276).Run: cluster.RunE: context canceled
```
<p>Parameters: <code>ROACHTEST_arch=amd64</code>
, <code>ROACHTEST_cloud=gce</code>
, <code>ROACHTEST_cpu=2</code>
, <code>ROACHTEST_encrypted=false</code>
, <code>ROACHTEST_fs=ext4</code>
, <code>ROACHTEST_localSSD=false</code>
, <code>ROACHTEST_ssd=0</code>
</p>
<details><summary>Help</summary>
<p>

See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)

See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*failover/chaos/read-write.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

Jira issue: CRDB-28685

Epic CRDB-27234

### LINKED ISSUE #105797: kvserver: config change proposals can hang forever
In https://github.com/cockroachdb/cockroach/issues/104709#issuecomment-1609474175 we see a replica circuit breaker continually (un)tripping on a stuck proposal.

During an `ALTER TABLE SCATTER`, three replicas on n6 trip circuit breakers because of a slow proposal. In all cases, the proposals are `EndTxn(commit change-replicas)`, i.e. they are attempting to commit an conf change, and all of the replicas are followers.

```
E230627 12:51:40.782735 148856 kv/kvserver/replica_circuit_breaker.go:175 ⋮ [T1,n6,s6,r1039/7:‹/Table/106/1/65{32837…-51266…}›] 30842 breaker: tripped with error: replica unavailable: (n6,s6):7 unable to serve request to r1039:‹/Table/106/1/65{32837936193842192-51266251951793792}› [(n3,s3):13, (n6,s6):7, (n8,s8):11, (n4,s4):14, (n9,s9):5, next=15, gen=92, sticky=9223372036.854775807,2147483647]: closed timestamp: 1687870297.597281357,0 (2023-06-27 12:51:37); raft status: {"id":"7","term":12,"vote":"e","commit":135,"lead":"e","raftState":"StateFollower","applied":135,"progress":{},"leadtransferee":"0"}: have been waiting 63.00s for slow proposal EndTxn(commit change-replicas) [/Local/Range/Table/106/1/‹6532837936193842192›/‹RangeDescriptor›], [txn: 494e1e1e]
E230627 12:51:43.283167 149136 kv/kvserver/replica_circuit_breaker.go:175 ⋮ [T1,n6,s6,r307/6:‹/Table/106/1/-68{2769…-0926…}›] 30847 breaker: tripped with error: replica unavailable: (n6,s6):6 unable to serve request to r307:‹/Table/106/1/-68{27690988321067808-09262672563116208}› [(n9,s9):7, (n7,s7):8, (n4,s4):4, (n3,s3):11, (n6,s6):6, next=12, gen=52, sticky=9223372036.854775807,2147483647]: closed timestamp: 1687870300.196827941,0 (2023-06-27 12:51:40); raft status: {"id":"6","term":12,"vote":"b","commit":127,"lead":"b","raftState":"StateFollower","applied":127,"progress":{},"leadtransferee":"0"}: have been waiting 60.50s for slow proposal EndTxn(commit change-replicas) [/Local/Range/Table/106/1/‹-6827690988321067808›/‹RangeDescriptor›], [txn: 321479c5]
E230627 12:51:44.282776 149410 kv/kvserver/replica_circuit_breaker.go:175 ⋮ [T1,n6,s6,r885/6:‹/Table/106/1/53{16569…-34997…}›] 30854 breaker: tripped with error: replica unavailable: (n6,s6):6 unable to serve request to r885:‹/Table/106/1/53{16569096169036592-34997411926988192}› [(n8,s8):11, (n4,s4):9, (n7,s7):12, (n5,s5):4, (n6,s6):6, next=13, gen=68, sticky=9223372036.854775807,2147483647]: closed timestamp: 1687870301.210269924,0 (2023-06-27 12:51:41); raft status: {"id":"6","term":13,"vote":"9","commit":122,"lead":"9","raftState":"StateFollower","applied":122,"progress":{},"leadtransferee":"0"}: have been waiting 61.50s for slow proposal EndTxn(commit change-replicas) [/Local/Range/Table/106/1/‹5316569096169036592›/‹RangeDescriptor›], [txn: 336523e8]
```

After being tripped, we continually see these replicas re-tripping the circuit breakers every 3 seconds, as they try to repropose these EndTxns. However, the Raft groups are fully functional, and the circuit breakers also reset immediately, but keep re-tripping on these proposals that don't appear to be going through.

Screenshot 2023-06-29 at 07 55 12

We even see it after n6 regains leadership of the Raft group:

```
E230627 13:08:34.784088 650606 kv/kvserver/replica_circuit_breaker.go:175 ⋮ [T1,n6,s6,r1039/7:‹/Table/106/1/65{32837…-51266…}›] 34156 breaker: tripped with error: replica unavailable: (n6,s6):7 unable to serve request to r1039:‹/Table/106/1/65{32837936193842192-51266251951793792}› [(n3,s3):13, (n6,s6):7, (n8,s8):11, (n4,s4):14, (n9,s9):5, next=15, gen=92, sticky=9223372036.854775807,2147483647]: closed timestamp: 1687871114.188750889,0 (2023-06-27 13:05:14); raft status: {"id":"7","term":13,"vote":"7","commit":963,"lead":"7","raftState":"StateLeader","applied":963,"progress":{"5":{"match":963,"next":965,"state":"StateReplicate"},"7":{"match":963,"next":964,"state":"StateReplicate"},"b":{"match":963,"next":965,"state":"StateReplicate"},"d":{"match":963,"next":965,"state":"StateReplicate"},"e":{"match":963,"next":965,"state":"StateReplicate"}},"leadtransferee":"0"}: have been waiting 1077.00s for slow proposal EndTxn(commit change-replicas) [/Local/Range/Table/106/1/‹6532837936193842192›/‹RangeDescriptor›], [txn: 494e1e1e]
```

We also see the transactions being cancelled:

```
I230627 12:51:40.785010 124952 kv/txn.go:878 ⋮ [T1,n6] 30843 async rollback failed: replica unavailable: (n6,s6):7 unable to serve request to r1039:‹/Table/106/1/65{32837936193842192-51266251951793792}› [(n3,s3):13, (n6,s6):7, (n8,s8):11, (n4,s4):14, (n9,s9):5, next=15, gen=92, sticky=9223372036.854775807,2147483647]: closed timestamp: 1687870297.597281357,0 (2023-06-27 12:51:37); raft status: {"id":"7","term":12,"vote":"e","commit":135,"lead":"e","raftState":"StateFollower","applied":135,"progress":{},"leadtransferee":"0"}: encountered poisoned latch ‹/Meta2/Table/106/1/6551266251951793792›@1687870207.598066812,0: "change-replica" meta={id=494e1e1e key=/Local/Range/Table/106/1/‹6532837936193842192›/‹RangeDescriptor› iso=Serializable pri=0.06378614 epo=0 ts=1687870207.598066812,0 min=1687870207.598066812,0 seq=2} lock=true stat=PENDING rts=1687870207.598066812,0 wto=false gul=1687870208.098066812,0
I230627 12:51:43.285782 124887 kv/txn.go:878 ⋮ [T1,n6] 30848 async rollback failed: replica unavailable: (n6,s6):6 unable to serve request to r307:‹/Table/106/1/-68{27690988321067808-09262672563116208}› [(n9,s9):7, (n7,s7):8, (n4,s4):4, (n3,s3):11, (n6,s6):6, next=12, gen=52, sticky=9223372036.854775807,2147483647]: closed timestamp: 1687870300.196827941,0 (2023-06-27 12:51:40); raft status: {"id":"6","term":12,"vote":"b","commit":127,"lead":"b","raftState":"StateFollower","applied":127,"progress":{},"leadtransferee":"0"}: encountered poisoned latch ‹/Meta2/Table/106/1/-6809262672563116208›@1687870210.148120953,0: "change-replica" meta={id=321479c5 key=/Local/Range/Table/106/1/‹-6827690988321067808›/‹RangeDescriptor› iso=Serializable pri=0.02487111 epo=0 ts=1687870210.148120953,0 min=1687870210.148120953,0 seq=2} lock=true stat=PENDING rts=1687870210.148120953,0 wto=false gul=1687870210.648120953,0
I230627 12:51:44.285303 124854 kv/txn.go:878 ⋮ [T1,n6] 30855 async rollback failed: replica unavailable: (n6,s6):6 unable to serve request to r885:‹/Table/106/1/53{16569096169036592-34997411926988192}› [(n8,s8):11, (n4,s4):9, (n7,s7):12, (n5,s5):4, (n6,s6):6, next=13, gen=68, sticky=9223372036.854775807,2147483647]: closed timestamp: 1687870301.210269924,0 (2023-06-27 12:51:41); raft status: {"id":"6","term":13,"vote":"9","commit":122,"lead":"9","raftState":"StateFollower","applied":122,"progress":{},"leadtransferee":"0"}: encountered poisoned latch ‹/Meta2/Table/106/1/5334997411926988192›@1687870210.000120522,0: "change-replica" meta={id=336523e8 key=/Local/Range/Table/106/1/‹5316569096169036592›/‹RangeDescriptor› iso=Serializable pri=0.00570728 epo=0 ts=1687870210.000120522,0 min=1687870210.000120522,0 seq=2} lock=true stat=PENDING rts=1687870210.000120522,0 wto=false gul=1687870210.500120522,0
```

Jira issue: CRDB-29202

### PR DESCRIPTION
**Pull request by tbg**
Because etcd/raft activates configuration changes when they are applied, but checks new proposed configs before they are considered for adding to the log (or forwarding to the leader), the following can happen:

- conf change 1 gets evaluated on a leaseholder n1
- lease changes
- new leaseholder evaluates and commits conf change 2
- n1 receives and applies conf change 2
- conf change 1 gets added to the proposal buffer and flushed; RawNode rejects
  it because conf change 1 is not compatible on top of conf change 2

Prior to this commit, because raft silently replaces the conf change with an
empty entry, we would never see the proposal below raft (where it would be
rejected due to the lease change). In effect, this caused replica unavailability
because the proposal and the associated latch would stick around forever, and the
replica circuit breaker would trip.

This commit provides a targeted fix: when the proposal buffer flushes a conf
change to raft, we check if it got replaced with an empty entry. If so, we
properly finish the proposal. To be conservative, we signal it with an ambiguous
result: it seems conceivable that the rejection would only occur on a
reproposal, while the original proposal made it into raft before the lease
change, and the local replica is in fact behind on conf changes rather than
ahead (which can happen if it's a follower). The only "customer" here is the
replicate queue (and scatter, etc) so this is acceptable; any choice here would
necessarily be a "hard error" anyway.

Epic: CRDB-25287
Release note (bug fix): under rare circumstances, a replication change could get
stuck when proposed near lease/leadership changes (and likely under overload),
and the replica circuit breakers could trip. This problem has been addressed.

Fixes #105797.
Closes #104709.

### CODE REVIEW

File: pkg/kv/kvserver/replica_proposal_buf.go [hunk 1]
```diff
@@ -563,16 +567,47 @@ func (b *propBuf) FlushLockedWithRaftGroup(
...
+			if err != nil {
+				firstErr = err
+				continue
+			}
+			msg := raftpb.Message{Type: raftpb.MsgProp, Entries: []raftpb.Entry{
+				{
+					Type: typ,
+					Data: data,
+				},
+			}}
```
Comment (erikgrinaker) [inline pkg/kv/kvserver/replica_proposal_buf.go:580]:
```suggestion
			msg := raftpb.Message{Type: raftpb.MsgProp, Entries: []raftpb.Entry{
				{Type: typ, Data: data},
			}}
```

File: pkg/kv/kvserver/client_raft_test.go [hunk 2]
```diff
@@ -6933,3 +6934,62 @@ func TestStoreMetricsOnIncomingOutgoingMsg(t *testing.T) {
...
+	tc := testcluster.StartTestCluster(t, 1, base.TestClusterArgs{ReplicationMode: base.ReplicationManual})
+	defer tc.Stopper().Stop(ctx)
+
+	k := tc.ScratchRange(t)
+
+	repl := tc.GetFirstStoreFromServer(t, 0).LookupReplica(keys.MustAddr(k))
+
+	// Try to leave a joint config even though we're not in one. This is something
+	// that will lead raft to propose an empty entry instead of our conf change.
+	// See:
```
Comment (erikgrinaker) [inline pkg/kv/kvserver/client_raft_test.go:6960]: Missing reference.

**Review by erikgrinaker — APPROVED** [review]
Great with the targeted fix!

**Comment by tbg** [comment]
TFTR!

bors r=erikgrinaker

**Comment by nvanbenschoten** [comment]
I just saw kvnemesis run into this new `config change rejected by raft; please retry` error when trying to reproduce https://github.com/cockroachdb/cockroach/issues/101721. Do we want to add retry handing for this error?

**Comment by tbg** [comment]
Yep let me look at it, thanks for the heads up!

**Comment by tbg** [comment]
Reopening, since we had to revert the fix: https://github.com/cockroachdb/cockroach/pull/106267

**Pull request merged**